File: src/main/java/com/moira/itda/global/auth/JwtProvider.java
package com.moira.itda.global.auth;

import com.moira.itda.global.entity.User;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.util.Base64;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

@Component
public class JwtProvider {
    @Value("${jwt.secret.key}")
    private String secretKey;

    private static final String JWT_ISSUER = "ITDA_LOCAL";                    // 토큰 발급자
    private static final Long JWT_EXPIRATION_TIME_ATK = 1000 * 60 * 60L;      // 1시간
    private static final Long JWT_EXPIRATION_TIME_RTK = 1000 * 60 * 60 * 24L; // 24시간
    private SecretKey key;

    @PostConstruct
    void init() {
        this.key = Keys.hmacShaKeyFor(
                Base64.getDecoder().decode(secretKey)
        );
    }

    public String createAtk(User user) {
        return this.createToken(user, JWT_EXPIRATION_TIME_ATK);
    }

    public String createRtk(User user) {
        return this.createToken(user, JWT_EXPIRATION_TIME_RTK);
    }

    private String createToken(User user, Long expirationTime) {
        Date now = new Date();
        Map<String, Object> claims = new HashMap<String, Object>();

        claims.put("email", user.getEmail());
        claims.put("nickname", user.getNickname());
        claims.put("role", user.getRole());

        return Jwts.builder()
                .subject(user.getId())
                .claims(claims)
                .expiration(new Date(now.getTime() + expirationTime))
                .issuedAt(now)
                .issuer(JWT_ISSUER)
                .signWith(key)
                .compact();
    }

    public String substringToken(String value) {
        return value.substring("Bearer ".length());
    }

    public Claims validate(String atk) {
        return Jwts.parser().verifyWith(key).build().parseSignedClaims(atk).getPayload();
    }
}

File: src/main/kotlin/com/moira/itda/global/auth/component/JwtProvider.kt
package com.moira.itda.global.auth.component

import com.moira.itda.global.auth.dto.UserAuth
import com.moira.itda.global.entity.User
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import jakarta.annotation.PostConstruct
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.*
import javax.crypto.SecretKey

@Component
class JwtProvider(
    @Value("\${jwt.expiration-time.atk}")
    private val expirationTimeAtk: Long,
    @Value("\${jwt.expiration-time.rtk}")
    private val expirationTimeRtk: Long
) {
    @Value("\${jwt.secret.key}")
    private lateinit var secretKey: String

    private lateinit var key: SecretKey

    /**
     * SecretKey 정의
     */
    @PostConstruct
    fun init() {
        this.key = Keys.hmacShaKeyFor(Base64.getDecoder().decode(secretKey))
    }

    /**
     * 토큰 생성
     */
    fun createToken(user: User, expirationTime: Long): String {
        val now = Date()
        val claims = HashMap<String, Any>().let {
            it["email"] = user.email
            it["nickname"] = user.nickname
            it["role"] = user.role.name
            it
        }

        return Jwts.builder()
            .subject(user.id)
            .claims(claims)
            .expiration(Date(now.time + expirationTime))
            .issuedAt(now)
            .issuer("ITDA")
            .signWith(key)
            .compact()
    }

    /**
     * AccessToken 생성
     */
    fun createAtk(user: User): String {
        return this.createToken(user, expirationTimeAtk)
    }

    /**
     * RefreshToken 생성
     */
    fun createRtk(user: User): String {
        return this.createToken(user, expirationTimeRtk)
    }

    /**
     * 토큰 substring (Bearer 제거)
     */
    fun substringToken(value: String): String {
        return value.substring("Bearer ".length)
    }

    /**
     * 토큰 검증
     */
    fun validateToken(token: String): Result<Claims?> {
        return runCatching {
            Jwts.parser().verifyWith(key).build().parseSignedClaims(token).payload
        }
    }

    /**
     * Claims에서 사용자 정보 추출
     */
    fun getUserAuth(claims: Claims): UserAuth {
        val userId = claims.subject
        val email = claims["email"] as String
        val nickname = claims["nickname"] as String
        val role = claims["role"] as String

        return UserAuth(userId = userId, email = email, nickname = nickname, role = role)
    }
}
